test-util/src/main/java/io/camunda/zeebe/test/util/bpmn/random/blocks/BoundaryEventBuilder.java

```diff
@@ -9,31 +9,67 @@
 
 import io.camunda.zeebe.model.bpmn.builder.AbstractActivityBuilder;
 import io.camunda.zeebe.model.bpmn.builder.AbstractFlowNodeBuilder;
-import io.camunda.zeebe.model.bpmn.builder.ExclusiveGatewayBuilder;
+import io.camunda.zeebe.test.util.bpmn.random.ConstructionContext;
+import io.camunda.zeebe.test.util.bpmn.random.RandomProcessGenerator;
 
 public class BoundaryEventBuilder {
 
   private final String taskId;
   private final String joinGatewayId;
-  private final ExclusiveGatewayBuilder exclusiveGatewayBuilder;
+  private boolean joinGatewayCreated;
+  private final boolean timerEventHasTerminateEndEvent;
+  private final boolean errorEventHasTerminateEndEvent;
 
-  public BoundaryEventBuilder(
-      final String taskId, final AbstractActivityBuilder<?, ?> taskBuilder) {
+  public BoundaryEventBuilder(final ConstructionContext context, final String taskId) {
     this.taskId = taskId;
     joinGatewayId = "boundary_join_" + taskId;
-    exclusiveGatewayBuilder = taskBuilder.exclusiveGateway(joinGatewayId);
+    timerEventHasTerminateEndEvent =
+        context.getRandom().nextDouble() < RandomProcessGenerator.PROBABILITY_TERMINATE_END_EVENT;
+    errorEventHasTerminateEndEvent =
+        context.getRandom().nextDouble() < RandomProcessGenerator.PROBABILITY_TERMINATE_END_EVENT;
+  }
+
+  private AbstractFlowNodeBuilder<?, ?> connectJoinGateway(
+      final AbstractFlowNodeBuilder<?, ?> taskBuilder) {
+    joinGatewayCreated = true;
+    return taskBuilder.exclusiveGateway(joinGatewayId);
   }
 
   public AbstractFlowNodeBuilder<?, ?> connectBoundaryErrorEvent(
-      final String boundaryEventId, final String errorCode) {
-    return ((AbstractActivityBuilder<?, ?>) exclusiveGatewayBuilder.moveToNode(taskId))
-        .boundaryEvent(boundaryEventId, b -> b.error(errorCode))
-        .connectTo(joinGatewayId);
+      final AbstractFlowNodeBuilder<?, ?> taskBuilder,
+      final String errorCode,
+      final String boundaryEventId) {
+    var builder = taskBuilder;
+    if (!joinGatewayCreated) {
+      builder = connectJoinGateway(taskBuilder);
```

Comment:
You misunderstood what this code does. The `connectJoinGateway` method connects the task to the gateway. So it's responsible for this blue part of the process.
<img width="425" alt="image" src="https://user-images.githubusercontent.com/5787702/202414182-2491456e-cf42-4aba-8f3d-2701a462cc0f.png">

We don't want to have 2 gateways, so we keep track if we have created this gateway already. If we did we don't connect it another time.

Later in the code we connect the boundary event to this gateway:
```java
    if (errorEventHasTerminateEndEvent) {
      return boundaryEventBuilder.endEvent().terminate().moveToNode(flowNodeId);
    } else {
      return boundaryEventBuilder.connectTo(joinGatewayId);
    }
```

Reply:
👍 Hmm. I see, this used to happen in the constructor of the boundary event block builder, but is now moved into the specific connect X boundary event methods. It's clear to me now.

🔧 Let's add an additional condition to the `if(joinGatewayCreated)` statement, to only do this when the boundary event is going to be connected to gateway. Or in other words, when `! errorEventHasTerminateEndEvent` for the error event; and when `! timerEventHasTerminateEndEvent`, because the terminate end event will not be connected to the gateway. Instead, the task should be directly connected to the next element in the sequence.